apps/bot_manager/lib/game_socket_handler.ex

```diff
@@ -123,6 +136,74 @@ defmodule BotManager.GameSocketHandler do
     {:ok, update_block_attack_state(state)}
   end
 
+  defp maybe_build_map(%{can_build_map: false} = state), do: state
+
+  defp maybe_build_map(%{bot_state_machine: %{collision_grid: collision_grid}} = state) when not is_nil(collision_grid),
+    do: state
+
+  defp maybe_build_map(%{game_state: nil} = state), do: state
+  defp maybe_build_map(%{game_state: %{obstacles: nil}} = state), do: state
+  defp maybe_build_map(%{game_state: %{obstacles: []}} = state), do: state
+
+  defp maybe_build_map(state) do
+    obstacles =
+      state.game_state.obstacles
+      |> Enum.map(fn {obstacle_id, obstacle} ->
+        {obstacle_id,
+         Map.take(Map.from_struct(obstacle), [
+           :id,
+           :shape,
+           :position,
+           :radius,
+           :vertices,
+           :speed,
+           :category,
+           :direction,
+           :is_moving,
+           :name
+         ])}
+      end)
+      |> Enum.map(fn {obstacle_id, obstacle} ->
+        {obstacle_id, Map.put(obstacle, :position, %{x: obstacle.position.x, y: obstacle.position.y})}
+      end)
+      |> Enum.map(fn {obstacle_id, obstacle} ->
+        {obstacle_id,
+         Map.put(
+           obstacle,
+           :vertices,
+           Enum.map(obstacle.vertices.positions, fn position -> %{x: position.x, y: position.y} end)
+         )}
+      end)
+      |> Enum.map(fn {obstacle_id, obstacle} ->
+        {obstacle_id, Map.put(obstacle, :direction, %{x: obstacle.direction.x, y: obstacle.direction.y})}
+      end)
+      |> Enum.map(fn {obstacle_id, obstacle} ->
+        {obstacle_id, Map.put(obstacle, :shape, get_shape(obstacle.shape))}
+      end)
+      |> Enum.map(fn {obstacle_id, obstacle} ->
+        {obstacle_id, Map.put(obstacle, :category, get_category(obstacle.category))}
+      end)
+      |> Map.new()
```

Comment:
Can we just merge all these enum.maps into one? So we don't have to traverse the list N times being N the amount of changes needed.